src/vec_int64_ppc.h

```diff
@@ -92,14 +100,752 @@
+  m1 = (vui32_t) vec_xxspltd ((vui64_t) a, 0);
+
+  // Need the endian invariant multiply even/odd word here
+#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+  // Nullify the little endian transform
+  p1 = vec_muleuw (m1, m0);
+  p0 = vec_mulouw (m1, m0);
+#else
+  p0 = vec_muleuw (m1, m0);
+  p1 = vec_mulouw (m1, m0);
+#endif
+  // res[1] = p1[1];  res[0] = p0[0];
+  res = vec_pasted (p0, p1);
+
+  // pp10[1] = p1[0]; pp10[0] = 0;
+  // pp01[1] = p0[1]; pp01[0] = 0;
+  // Need the endian invariant merge algebraic high/low here
+  pp10 = (vui64_t) vec_mrgahd ((vui128_t) zero, (vui128_t) p1);
+  pp01 = (vui64_t) vec_mrgald ((vui128_t) zero, (vui128_t) p0);
+  // pp01 = pp01 + pp10.
+  pp01 = (vui64_t) vec_adduqm ((vui128_t) pp01, (vui128_t) pp10);
+
+  // res = res + (pp01 << 32)
+  pp01 = (vui64_t) vec_sld ((vi32_t) pp01, (vi32_t) pp01, 4);
+  res = (vui64_t) vec_adduqm ((vui128_t) pp01, (vui128_t) res);
+#else
+  // _ARCH_PWR7 implementation ...
+#endif
+#endif
+  return ((vui128_t) res);
+}
+ * \endcode
+ * The _ARCH_PWR9 implementation uses the multiply-sum doubleword
+ * operation but implements the multiply even behavior by forcing the
+ * contents of doubleword element 1 of [VRB] and the contents of
+ * [VRC] to 0.
+ *
+ * The _ARCH_PWR8
+ * implementation looks ugly but it works. It starts with some merges
+ * and splats to get inputs columns lined up for the multiply.
+ * Then we use (POWER8 instructions) Multiply even/odd unsigned word
+ * to generate doubleword partial products. Then more
+ * merges and a rotate to line up the partial products for summation
+ * as the final quadword product.
+ *
+ * Individually vec_vmuleud() and vec_vmuloud() execute with a latency
+ * of 21-23 cycles on POWER8. Normally these operations are used and
+ * scheduled together as in the POWER8 implementation of vec_msumudm()
+ * or vec_mulhud(). Good scheduling by the compiler and pipelining
+ * keeps the POWER8 latency in the 28-32 cycle range.
+ * For example the vec_mulhud() implementation:
+ * \code
+static inline vui64_t
+vec_mulhud (vui64_t vra, vui64_t vrb)
+{
+  return vec_mrgahd (vec_vmuleud (vra, vrb), vec_vmuloud (vra, vrb));
+}
+ * \endcode
+ * Generates the following code for POWER8:
+ * \code
+	vspltisw v0,0
+	xxmrghw vs33,vs35,vs35
+	xxspltd vs45,vs34,0
+	xxmrglw vs35,vs35,vs35
+	vmulouw v11,v13,v1
+	xxspltd vs34,vs34,1
+	xxmrghd vs41,vs32,vs43
+	vmulouw v12,v2,v3
+	vmuleuw v13,v13,v1
+	vmuleuw v2,v2,v3
+	xxmrghd vs42,vs32,vs44
+	xxmrgld vs33,vs32,vs45
+	xxmrgld vs32,vs32,vs34
+	xxpermdi vs44,vs34,vs44,1
+	vadduqm v1,v1,v9
+	xxpermdi vs45,vs45,vs43,1
+	vadduqm v0,v0,v10
+	vsldoi  v1,v1,v1,4
+	vsldoi  v0,v0,v0,4
+	vadduqm v2,v1,v13
+	vadduqm v0,v0,v12
+	xxmrghd vs34,vs34,vs32
+ * \endcode
+ *
+ * The POWER9 latencies for this operations range from 5-7 (for
+ * vmsumudm itself) to 11-16 (for vec_mulhud()).
+ * The additional latency reflects generate of zero constant vector
+ * and merges requires to condition the inputs and output. For these
```

Comment:
The additional latency reflects zero constant vector generation and merges required to condition the inputs and output.

Reply:
OK